samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,751 @@
+	       ctx->ota_range_end_byte);
+
+	/* Check if firmware update is completed */
+	if (status == SL_STATUS_SI91X_FW_UPDATE_DONE) {
+		return 1;
+	} else if (status != SL_STATUS_OK) {
+		return -EIO;
+	}
+	return 0;
+}
+
+/*
+ * @brief Handles retry logic for OTA operations
+ *
+ * @param ctx Pointer to application context
+ * @param operation_name Name of the operation being retried for logging
+ * @param cleanup_socket Whether to clean up socket resources before retrying
+ * @return true if retry should continue, false if max retries exceeded
+ */
+static bool ota_handle_retry(struct app_ctx *ctx, const char *operation_name, bool cleanup_socket)
+{
+	/* Clean up socket if needed */
+	if (cleanup_socket) {
+		ota_cleanup_resources(ctx);
+	}
+	if (++ctx->retry_count > 2) {
+		printf("Maximum retries exceeded, aborting OTA\n");
+		return false;
+	}
+	printf("Retrying %s (%d/%d)...\n", operation_name, ctx->retry_count, 2);
+	k_sleep(K_MSEC(1000));
+	return true;
+}
+
+/*
+ * @brief Main OTA application state machine
+ * @param ctx Pointer to application context
+ */
+static void ota_application_start(struct app_ctx *ctx)
+{
+	char *ssid = CONFIG_OTA_WIFI_SSID;
+	char *pwd = CONFIG_OTA_WIFI_PSK;
+	bool ota_complete = false;
+	const char *operation;
+	int ret = 0;
+
+	printf("OTA Application Started\n");
+	/* Main state machine loop */
+	while (ctx->retry_count < 3) {
+		switch (ctx->state) {
+		case OTA_STATE_SCAN:
+			/* Scan for networks and connect */
+			ota_wifi_start_scan(ctx);
+			ret = ota_wifi_connect(ssid, pwd, ctx);
+			if (ret) {
+				if (!ota_handle_retry(ctx, "Wi-Fi connection", false)) {
+					ctx->retry_count++;
+					continue;
+				}
+				break;
+			}
+
+			/* Wait for IP configuration */
+			ret = ota_wait_for_ip_config(ctx);
+			if (ret < 0) {
+				return;
+			}
+			ctx->retry_count = 0;
+
+			ctx->state = OTA_STATE_SERVER_CONNECT;
+			break;
+
+		case OTA_STATE_SERVER_CONNECT:
+			ret = ota_connect_to_server(ctx);
+			if (ret < 0) {
+				if (!ota_handle_retry(ctx, "server connection", true)) {
+					ctx->retry_count++;
+					continue;
+				}
+				break;
+			}
+			ctx->state = OTA_STATE_DOWNLOAD;
+			break;
+
+		case OTA_STATE_DOWNLOAD:
+			if (!ctx->ipv4_addr_config_done || !ctx->ipv6_addr_config_done ||
+			    ctx->disonnected) {
+				ctx->state = OTA_STATE_SCAN;
+				ota_cleanup_resources(ctx);
+				break;
+			}
+			/* Download firmware chunk */
+			ret = ota_download_firmware_chunk(ctx);
+			if (ret < 0) {
+				if (!ota_handle_retry(ctx, "download", true)) {
+					ctx->retry_count++;
+					continue;
+				}
+				ctx->state = OTA_STATE_SERVER_CONNECT;
+				break;
+			}
+
+			ret = ota_load_firmware(ctx);
+			if (ret < 0) {
+				if (ret == -EINVAL) {
+					ctx->state = OTA_STATE_SERVER_CONNECT;
+					operation = "download";
+
+				} else {
+					ctx->state = (ctx->sock < 0) ? OTA_STATE_SERVER_CONNECT
+								     : OTA_STATE_DOWNLOAD;
+					operation =
+						(ctx->sock < 0) ? "server connection" : "download";
```

Comment:
I believe `ctx->sock < 0` and `ctx->state == OTA_STATE_SERVER_CONNECT` are synonym, You could probably get rid of `OTA_STATE_SERVER_CONNECT` and only test `ctx->sock < 0`.